Bukkit/src/main/java/me/egg82/tcpp/api/trolls/AnvilTroll.java

```diff
@@ -45,73 +44,43 @@ public void start(CommandIssuer issuer) throws Exception {
             return;
         }
 
-        events.add(
-                BukkitEvents.subscribe(plugin, EntityChangeBlockEvent.class, EventPriority.LOW)
-                        .handler(this::entityChangeBlock)
-        );
-
-        // TODO: save blocks and replace
-        Location anvilLocation = player.getLocation();
-        for (int i = 0; i < 3; i++) {
+        // TODO: inventory support (restore not working with chests or shulker boxes)
+        anvilLocation = player.getLocation();
+        for (int i = 0; i < 4; i++) {
+            oldBlockData[i] = anvilLocation.getBlock().getBlockData();
             anvilLocation.getBlock().setType(Material.AIR, false);
             anvilLocation.add(0.0d, 1.0d, 0.0d);
         }
-        anvilLocation.add(0.0d, 1.0d, 0.0d);
+        oldBlockData[4] = anvilLocation.getBlock().getBlockData();
+        anvilLocation.getBlock().setType(Material.AIR, false);
 
         if (fallingBlockMethod != null) {
             try {
-                anvilID = ((FallingBlock) fallingBlockMethod.invoke(anvilLocation.getWorld(), anvilLocation, new MaterialData(Material.ANVIL))).getUniqueId();
+                FallingBlock anvil = (FallingBlock) fallingBlockMethod.invoke(anvilLocation.getWorld(), anvilLocation, Material.ANVIL.createBlockData());
+                anvil.setHurtEntities(true);
+                anvil.setDropItem(false);
             } catch (IllegalAccessException | InvocationTargetException ex) {
                 logger.error(ex.getMessage(), ex);
             }
         } else {
             anvilLocation.getBlock().setType(Material.ANVIL);
-            scheduleGetAnvil(anvilLocation, 40);
         }
+        anvilLocation.add(0d, -4d, 0d);
```

Comment:
Sure, this is more efficient, but only slightly. Also, I'd like to adjust this height later and it's easier to change one number than to remember to change two.